crypto/fipsmodule/digest/digest.c

```diff
@@ -218,19 +224,34 @@ int EVP_MD_CTX_reset(EVP_MD_CTX *ctx) {
 
 int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *engine) {
   if (ctx->digest != type) {
-    assert(type->ctx_size != 0);
-    uint8_t *md_data = OPENSSL_malloc(type->ctx_size);
-    if (md_data == NULL) {
-      return 0;
-    }
-
-    OPENSSL_free(ctx->md_data);
-    ctx->md_data = md_data;
     ctx->digest = type;
+    if (!used_for_hmac(ctx)) {
+      assert(type->ctx_size != 0);
+      ctx->update = type->update;
+      uint8_t *md_data = OPENSSL_malloc(type->ctx_size);
+      if (md_data == NULL) {
+        return 0;
+      }
+      OPENSSL_free(ctx->md_data);
+      ctx->md_data = md_data;
+    }
   }
 
   assert(ctx->pctx == NULL || ctx->pctx_ops != NULL);
```

Comment:
Not quite enough if `used_hmac` is true?

Reply:
I added an additional assertion within `used_for_hmac` below.

Reply:
I think what I had in mind was stronger test under the `used_hmac` block. Explicitly ensure some of the long dereference chains doesn't cause null-dereferencing.